src/archive.rs

```diff
@@ -0,0 +1,154 @@
+}
+
+/// Helper to open a tarfile for reading/writing.
+pub fn create_tarfile<P: AsRef<Utf8Path> + std::fmt::Debug>(tarfile: P) -> Result<File> {
+    OpenOptions::new()
+        .write(true)
+        .read(true)
+        .truncate(true)
+        .create(true)
+        .open(tarfile.as_ref())
+        .map_err(|err| anyhow!("Cannot create tarfile {:?}: {}", tarfile, err))
+}
+
+/// Helper to open a tarfile for reading.
+pub fn open_tarfile<P: AsRef<Utf8Path> + std::fmt::Debug>(tarfile: P) -> Result<File> {
+    OpenOptions::new()
+        .read(true)
+        .open(tarfile.as_ref())
+        .map_err(|err| anyhow!("Cannot open tarfile {:?}: {}", tarfile, err))
+}
+
+pub trait Encoder: std::io::Write + Send {}
+impl<T> Encoder for T where T: std::io::Write + Send {}
+
+pub struct ArchiveBuilder<E: Encoder> {
+    pub builder: tar::Builder<E>,
+}
+
+impl<E: Encoder> ArchiveBuilder<E> {
+    pub fn new(builder: tar::Builder<E>) -> Self {
+        Self { builder }
+    }
+
+    pub fn into_inner(self) -> Result<E> {
+        self.builder
+            .into_inner()
+            .with_context(|| "Finalizing archive")
+    }
+}
+
+/// Adds a package at `package_path` to a new zone image
+/// being built using the `archive` builder.
+pub async fn add_package_to_zone_archive<E: Encoder>(
+    archive: &mut ArchiveBuilder<E>,
+    package_path: &Utf8Path,
+) -> Result<()> {
+    let tmp = camino_tempfile::tempdir()?;
+    let gzr = flate2::read::GzDecoder::new(open_tarfile(package_path)?);
+    if gzr.header().is_none() {
+        return Err(anyhow!(
+            "Missing gzip header from {} - cannot add it to zone image",
+            package_path,
+        ));
+    }
+    let mut component_reader = tar::Archive::new(gzr);
+    let entries = component_reader.entries()?;
+
+    // First, unpack the existing entries
+    for entry in entries {
+        let mut entry = entry?;
+
+        // Ignore the JSON header files
+        let entry_path = entry.path()?;
+        if entry_path == Utf8Path::new("oxide.json") {
+            continue;
+        }
+
+        let entry_path: &Utf8Path = entry_path.strip_prefix("root/")?.try_into()?;
+        let entry_unpack_path = tmp.path().join(entry_path);
+        entry.unpack(&entry_unpack_path)?;
+
+        let entry_path = entry.path()?.into_owned();
+        let entry_path: &Utf8Path = entry_path.as_path().try_into()?;
+        assert!(entry_unpack_path.exists());
+
+        archive
+            .builder
+            .append_path_with_name_async(entry_unpack_path, entry_path)
+            .await?;
+    }
+    Ok(())
+}
+
+pub async fn new_compressed_archive_builder(
+    path: &Utf8Path,
+) -> Result<ArchiveBuilder<GzEncoder<File>>> {
+    let file = create_tarfile(path)?;
+    // TODO: Consider using async compression, async tar.
+    // It's not the *worst* thing in the world for a packaging tool to block
+    // here, but it would help the other async threads remain responsive if
+    // we avoided blocking.
```

Comment:
👍 full agreement that this isn't needed but could be cool (I know this comment was moved from old code, but imho it's still true and so worth remaking on)

Reply:
Totally! With the `block_on` calls this is slightly mitigated, but it still has a small cost.